### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by ahelal**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,91 @@
...
+
+#### `out Parameters`
+
+* `path`: *Required.* The path of the resource name. To fetch original message ts
+
+* `reply`: *Required.* The content of the reply. This can be also a template.
+
+### Template
+
+Under the hood the template use python [Jinj2](http://jinja.pocoo.org/docs/2.9/).
```
Comment (aleerizw-zz) [inline README.md]: Should be jinja2

File: README.md [hunk 2]
```diff
@@ -1,2 +1,91 @@
...
+
+* `variables`: You can access variables using curly braces `{{ VARIABLE_NAME }}`.
+   All environmental variables are accessible through `{{ ENV['PATH'] }}`. Concourse exposes some metadata info like job_name, build and ... for more info [check concourse website](https://concourse.ci/implementing-resources.html#resource-metadata)
+   The regex groups are accessible to the template engine as `{{ regex }}` if you used *subgroups* in your expression you can access each group with index `{{ regex[0] }}`. If you used *named subgroups* you can access them as dictionary `{{ regex['name'] }}`.
+
+* `white spaces`: You can use `\n` for new lines `\t`.
```
Comment (aleerizw-zz) [inline README.md]: Doesn't make sense

File: README.md [hunk 3]
```diff
@@ -1,2 +1,91 @@
-# bender
-A Slack concourse CI resource that can trigger jobs
+# bender Resource
+
+A  Concourse resource that can trigger any job (deployments, releases, ...) using slack.
+
+## Source Configuration
+
+* `slack_token`: *Required.* The slack token to be used. For more info check [slack](https://api.slack.com/docs/token-types#bot)
```
Comment (rnurgaliyev) [inline README.md]: Maybe add 'documentation' and the end of the sentence.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,91 @@
...
+
+* `channel`: *Required.* The channel name to be used by the bot.
+
+* `bot_name`: *Optional.*, *default `bender`* The bot name will be used to identify and filter messages. All messages must be addressed to the bot `@bot_name some message`.
```
Comment (rnurgaliyev) [inline README.md]: *Optional.*, *default `bender`* doesn't read well. Maybe add commas and stops.

Comment (rnurgaliyev) [reply README.md]: All messages must be addressed to the bot `@bot_name some message`. doesn't read well. Maybe redo this sentence like that: All messages must be addressed to the bot, e.g.: `@bot_name some message`.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,91 @@
...
+
+* `template`: *Optional.* A string that will be evaluated and written to `template_filename` can be used as an input file for further jobs in the pipeline.
+
+* `template_filename`: *Optional.* *default `template_file.txt`* the file name for a generated template.
+
+## Behavior
+
+### `check`: Check for new messages that match the rules
+
+Check will report a new version `timestamp` if messages
```
Comment (rnurgaliyev) [inline README.md]: Semicolon?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,91 @@
...
+
+1: Are addressed to `@bot_name`.
+
+2: If `grammar` is defined it matches the regex of the message.
```
Comment (rnurgaliyev) [inline README.md]: I don't get meaning in this sentence. Maybe redo it like this: Check will report new version if any message in bot channel matches regular expression defined in `grammar`.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,91 @@
...
+
+#### `check Parameters`
+
+Check accepts no params.
+
+### `in`: Get the message
+
+A file **bender.json** will be created with the message payload.
+
+if `template` is defined it will be evaluated and written `template_filename`. For more info on syntax check [Template section](#Template) for example usage check the [examples](#Examples)
```
Comment (rnurgaliyev) [inline README.md]: and written **to**, please add comas.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,91 @@
...
+
+#### `in Parameters`
+
+In accepts no params.
+
+### `out`: Reply to original message thread
```
Comment (rnurgaliyev) [inline README.md]: `Reply to original message` or `reply in original thread` reads better.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,91 @@
...
+
+Sends a message as a thread to the original message sent by the user.
```
Comment (rnurgaliyev) [inline README.md]: maybe `Replies to original message in a thread.`

File: README.md (hunk 2 above)
Comment (rnurgaliyev) [inline README.md]: I don't get the sentence. Possibly comas are missing.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,91 @@
...
+
+Under the hood the template use python [Jinj2](http://jinja.pocoo.org/docs/2.9/).
+
+* `variables`: You can access variables using curly braces `{{ VARIABLE_NAME }}`.
+   All environmental variables are accessible through `{{ ENV['PATH'] }}`. Concourse exposes some metadata info like job_name, build and ... for more info [check concourse website](https://concourse.ci/implementing-resources.html#resource-metadata)
+   The regex groups are accessible to the template engine as `{{ regex }}` if you used *subgroups* in your expression you can access each group with index `{{ regex[0] }}`. If you used *named subgroups* you can access them as dictionary `{{ regex['name'] }}`.
+
+* `white spaces`: You can use `\n` for new lines `\t`.
+
+* `filters`: You can use Jinja2 [built filters](http://jinja.pocoo.org/docs/2.9/templates/#builtin-filters).
```
Comment (rnurgaliyev) [inline README.md]: built**in** filters?

File: README.md [hunk 11]
```diff
@@ -1,2 +1,91 @@
...
+
+## Tips
+
+* Test your regex using something like [regexr.com](http://regexr.com/), It is better to use [named subgroups](http://www.regular-expressions.info/brackets.html).
```
Comment (rnurgaliyev) [inline README.md]: It->it

File: README.md [hunk 12]
```diff
@@ -1,2 +1,91 @@
...
+
+* Note that you need to escape `\` with another `\\` in your expression.
```
Comment (rnurgaliyev) [inline README.md]: Maybe rephrase as double-escape.

Comment (rnurgaliyev) [reply README.md]: Because now it sounds like you need to escape `\` with `\\` which will make `\\\` \\\\\ \\\\\\\\ \\\ \\\\

File: README.md [hunk 13]
```diff
@@ -1,2 +1,91 @@
...
+
+* If you have multiple triggers on the same channel it is better to use a different bot token for each. check TODO
+
+## Examples
+
+Check [examples.md](examples.md)
+
+## TODO
+
+* At the moment we depend on `ts` to mark last message. So last read message is kept in slack rather then concourse native *check* version. The advantage of using slack it will be faster, but downside you can't have multiple triggers per channel with same token. this only effects the check method.
```
Comment (rnurgaliyev) [inline README.md]: this -> This, effects -> affects. `The advantage of using slack it will be faster` reads badly, maybe rephrase the sentence to smth like `The advantage of using slack is better speed`

File: README.md [hunk 14]
```diff
@@ -1,2 +1,91 @@
...
+
+* `reply`: *Required.* The content of the reply. This can be also a template.
+
+### Template
+
+Under the hood the template use python [Jinj2](http://jinja.pocoo.org/docs/2.9/).
+
+* `variables`: You can access variables using curly braces `{{ VARIABLE_NAME }}`.
+   All environmental variables are accessible through `{{ ENV['PATH'] }}`. Concourse exposes some metadata info like job_name, build and ... for more info [check concourse website](https://concourse.ci/implementing-resources.html#resource-metadata)
+   The regex groups are accessible to the template engine as `{{ regex }}` if you used *subgroups* in your expression you can access each group with index `{{ regex[0] }}`. If you used *named subgroups* you can access them as dictionary `{{ regex['name'] }}`.
```
Comment (rnurgaliyev) [inline README.md]: Please add comas.

File: README.md [hunk 15]
```diff
@@ -1,2 +1,91 @@
...
+
+### `out`: Reply to original message thread
+
+Sends a message as a thread to the original message sent by the user.
+
+#### `out Parameters`
+
+* `path`: *Required.* The path of the resource name. To fetch original message ts
+
+* `reply`: *Required.* The content of the reply. This can be also a template.
```
Comment (rnurgaliyev) [inline README.md]: can be also -> can also be

Comment (rnurgaliyev) [reply README.md]: `To fetch original message ts` this sentence looks broken.

Comment (rnurgaliyev) [inline README.md]: I'd change word `content` with something more self-explanatory, like `message in the reply`.

File: README.md (hunk 1 above)
Comment (rnurgaliyev) [inline README.md]: Jinj2 -> Jinja2, sentence `Under the hood the template use python` looks broken to me. I'd rephrase: `Template uses Python Jinja2 templating engine`

File: README.md [hunk 16]
```diff
@@ -1,2 +1,94 @@
-# bender
-A Slack concourse CI resource that can trigger jobs
+# bender Resource
+
+A  Concourse resource that can trigger any job (deployments, releases, ...) using slack.
+Check [examples](#Examples)
```
Comment (rnurgaliyev) [inline README.md:4]: Add coma or better, something like `Check examples for better understaing.`

File: README.md [hunk 17]
```diff
@@ -1,2 +1,94 @@
...
+
+## Source Configuration
+
+* `slack_token`: *Required*. The slack token to be used. For more info check [slack documentation](https://api.slack.com/docs/token-types#bot).
+
+* `channel`: *Required*. The channel name to be used by the bot.
+
+* `bot_name`: *Optional*, *default `bender`*. The bot name will be used to identify and filter messages. All messages must be addressed to the bot, eg.: `@bot_name some message`.
+
+* `grammar`: *Optional.* If not defined bender will respond to all mentions `@bot_name`. If grammar is defined bender will respond to messages matching the regex expression. Use [python regular expression](https://docs.python.org/2/library/re.html) syntax. See [examples](#Examples) for inspiration.
```
Comment (rnurgaliyev) [inline README.md]: pls add **only** in this sentence.

File: README.md [hunk 18]
```diff
@@ -1,2 +1,94 @@
...
+
+* `template`: *Optional*. A string that will be evaluated and written to `template_filename` can be used as an input file for further jobs in the pipeline.
```
Comment (rnurgaliyev) [inline README.md]: These two sentences must be split by `and` or by coma.

File: README.md [hunk 19]
```diff
@@ -1,2 +1,94 @@
...
+
+* `white spaces`: You can use `\n` for new lines `\t` for tabs in your template.
+
+* `filters`: You can use Jinja2 [builtin filters](http://jinja.pocoo.org/docs/2.9/templates/#builtin-filters).
+
+## Tips
+
+* Test your regex using something like [regexr.com](http://regexr.com/), it is better to use [named subgroups](http://www.regular-expressions.info/brackets.html).
+
+* Note that you need to escape `\` with another `\\` in your expression.
```
Comment (rnurgaliyev) [inline README.md]: You didn't fix this one :P

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,13 @@
+FROM alpine
```
Comment (c4pone) [inline Dockerfile:1]: why not use the official python alpine image

File: README.md [hunk 21]
```diff
@@ -1,2 +1,94 @@
...
+
+#### `check Parameters`
+
+Check accepts no params.
+
+### `in`: Get the message
+
+A file **bender.json** will be created with the message payload.
+
+if `template` is defined it will be evaluated and written to, `template_filename`. For more info on syntax check [Template section](#Template) for example usage check the [examples](#Examples)
```
Comment (rnurgaliyev) [inline README.md]: Coma here is redundant.

Comment (rnurgaliyev) [reply README.md]: `For more info on syntax check read [Template section](#Template). For example on usage, check the [examples](#Examples)` section.`

File: requirments.txt [hunk 22]
```diff
@@ -0,0 +1,2 @@
+slackclient
+jinja2
```
Comment (c4pone) [inline requirments.txt:2]: better define a specific version here

File: src/lib/base.py [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+    if not condition:
+        print("{}".format(msg), file=sys.stderr)
+        exit(1)
+
+
+class Base(object):
+    """Slack Concourse resource implementation"""
+
+    def __init__(self, **kwargs):
+        self.slack_client = SlackClient(kwargs.get("slack_token"))
```
Comment (c4pone) [inline src/lib/base.py:52]: instead of initialising the slack client directly here, it would be better to inject the dependency. Doing so, would make it easier to test it
